src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+{-# OPTIONS --safe #-}
+-- {-# OPTIONS -v tactic.premises:100 #-}
+module Tactic.Premises where
+
+open import Prelude
+import Data.List as L
+open import Data.Fin using (toℕ)
+
+open import Class.Show.Core using (show; Show)
+open import Class.Show.Instances using ()
+open import Class.Semigroup.Core; open import Class.Semigroup.Instances
+open import Class.Monad.Core; open import Class.Monad.Instances
+open import Class.Applicative.Core; open import Class.Applicative.Instances
+
+open import Interface.Functor using (map; _<$>_)
+open import Interface.Bifunctor using (map₁; map₂)
+open import Interface.DecEq.Ext
+open import Interface.Decidable.Instance
+open import Interface.STS
+
+open import Generics; open Debug ("tactic.premises", 100)
+open import Generics.Utils.Ext
+open import Agda.Builtin.Reflection using (onlyReduceDefs; getInstances)
+open import Reflection hiding (_>>=_; _>>_; _≟_)
+open import Reflection.AST.Term hiding (_≟_)
+open import Reflection.AST.Abstraction using (unAbs)
+
+{-
+** Extracting the hypotheses of an STS rule.
+
+Constructors should be formulated in the STS syntax for inference rules, i.e.
+```
+  rule : ∀ <IMPLICIT_ARGUMENTS + LET_EXPRESSIONS>
+    ∙ <HYPOTHESIS#1>
+    ∙ <HYPOTHESIS#2>
+    ⋮
+    ∙ <HYPOTHESIS#n-1>
+    ∙ <HYPOTHESIS#n>
+      ────────────────────────────────
+      <CONCLUSION>
+```
+This meta-program then generates the type of the (decidable) hypotheses of the rule,
+as well as a proof that it is in fact decidable, i.e.
+```
+rule-premises : ∀ ⋯ → Set
+rule-premises ⋯ =
+  ∙ <HYPOTHESIS#1>
+  ∙ <HYPOTHESIS#3>
+  ⋮
+  ∙ <HYPOTHESIS#n-1>
+
+rule-premises? : ∀ ⋯ → Dec (rule-premises ⋯)
+rule-premises? = it
+```
+where ⋯ stands for the subset of implicit arguments that are relevant for the
+selected subset of decidable hypotheses.
+
+To be precise, we generate both definitions above under a single name, i.e.
+```
+unquoteDecl rule-premises = genPremises rule-premises (quote rule)
+-- rule-premises        : ∀ ⋯ → ∃ Dec
+-- rule-premises .proj₁ : ∀ ⋯ → Set
+-- rule-premises .proj₂ : ∀ ⋯ → Dec (rule-premises .proj₁ ⋯)
+```
+-}
+genPremises : Name → Name → TC ⊤
+genPremises f n = do
+  print $ "*** Generating premises for constructor:" <+> show n <+> "***"
+  ty ← reduceRuleSyntax =<< getType n                  -- (1)
+  let is , hs = breakImplicits $ viewTy ty .proj₁      -- (2)
+  hs′ ← map unbundleHypotheses                         -- (3)
+      $ extendContextTel (absTelescope is)
+      $ mapM reduceRuleSyntax hs
+  dhs ← forM hs′ $ λ h → (_, h) <$> isDecidable? is h  -- (4)
```

Comment:
```suggestion
  dhs ← forM (unbundleHypotheses hs)
           $ λ h → (_, h) <$> isDecidable? is h        -- (4)
```

I was confused about why `reduceRuleSyntax` was used here again, since all it does is to reduce away notation that was already reduced in `(1)`. This is indeed unnecessary, and you can simplify away the whole monad dance. If you prefer not merging it into step `(4)`, you can also do `let hs = unbundleHypotheses hs`.

Reply:
It turns out this is necessary for some reason I don't completely understand, but if you look at the generated types without the second `reduceRuleSyntax` you see an artifact of typechecking that grabs all the context (hence all variables will be considered relevant, breaking our Utxow rule for instance), namely there is a term `mkGeneraliseTel (var n) ... (var 0)` which somehow gets generated during typechecking in `src/full/Agda/TypeChecking/Generalize.createGenRecordType` if that helps.

Reply:
@UlfNorell any idea what these `mkGeneraliseTel` terms are doing there?

Reply:
I think it's this: agda/agda#6951

Reply:
Turns out it's a reported bug: https://github.com/agda/agda/issues/6951